Switch to TypeScript, break the API a bit

JustinBeckwith opened the pull request:
A few things happening in here:

1. Switched the project to TypeScript, including gts.  Module now has a d.ts. 
2. Broke the API a bit.  Removed the sync version because I can't be convinced it's actually a good idea.
3. Removed the factory pattern from the constructor because it was confusing and not really needed.
4. Added support for promise/async style, while keeping callbacks. 

Are folks ok with the changing the API? 
Also, curious what I should do for a changelog here with the breaking changes.

ofrobots commented on src/index.ts:
```diff
@@ -0,0 +1,50 @@
+ */
+export function getPem(
+    filename: string, callback?: (err: Error|null, pem: string|null) => void) {
+  return new Promise<string>((resolve, reject) => {
```
Don't allocate a promise if a callback was provided. Not only is an unnecessary allocation, if the promise rejects, a few potential things can happen, all bad:
* By default unhandled rejection warning on the console.
* If the user is using something like [hard-rejection](https://github.com/sindresorhus/hard-rejection), the application will crash.

JustinBeckwith replied:
Good feedback.  Here's the problem - if I try to conditionally return a Promise with TypeScript, the compiler notices that and will enforce null checking on return values from getPem().  Instead - I still return a Promise, but I only invoke the resolve or reject methods if a callback isn't passed.  Thoughts?

ofrobots replied:
Tough. Summoning @jinwoo, @matthewloring in case they have some ideas.

ofrobots replied:
Aha. This might be useful: http://www.typescriptlang.org/docs/handbook/functions.html#overloads

jinwoo replied:
Agreed that type overloading is the way to handle this.

But I'm wondering why do we have to support both callback and promise?

ofrobots commented on .gitignore:
```diff
@@ -1 +1,4 @@
 node_modules
+.DS_Store
+build/
+*.js
```
nit: newline

JustinBeckwith replied:
added

ofrobots commented on .travis.yml:
```diff
@@ -1,5 +1,5 @@
-  - "7"
+  - "4"
   - "6"
+  - "8"
```
nit: newline.

JustinBeckwith replied:
added

ofrobots commented on README.md:
````diff
@@ -12,35 +12,26 @@ npm install google-p12-pem
+### async/await style
+```js
+const gp12 = require('google-p12-pem');
+const pem = await gp12.getPem('/path/to/key.p12');
````
Is it possible to make the module evaluation value be the function? That is, I would like to be able to do
```js
const pem = await gp12('/path/to/key.p12');
```

To achieve this, you can use the following in your `index.ts`:

```ts
export = getPem;
```

JustinBeckwith replied:
This means I'm not really using es modules, right?  If I change this, I can't use `import` syntax in other places.

ofrobots replied:
Ack.

ofrobots commented on package.json:
```diff
@@ -1,20 +1,39 @@
+    "README.md",
+    "build/src"
+  ],
+  "types": "./build/src/index.d.ts",
```
nit: it would be good to have this next to `"main"`.

ofrobots commented on src/index.ts:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Convert a .p12 file to .pem string
+ * @param  {string}    filename The .p12 key filename.
```
drop the types from these annotations.

ofrobots commented on tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+  "exclude": [
+    "node_modules"
+  ]
+}
```
nit: newline; or perhaps I need to open a bug against `gts`?

JustinBeckwith replied:
Yup :) https://github.com/google/ts-style/issues/68

ofrobots commented:
As for the changelog, my suggestion is to use GitHub releases for the purpose of the changelog as well. See the releases page for `gts`: https://github.com/google/ts-style/releases

ofrobots commented:
While I don't have experience with it, https://www.npmjs.com/package/release can make it easier to easily generate good GitHub releases with changelogs.

JustinBeckwith commented:
As soon as it's time to generate a release, I can go ahead and do that.  @ofrobots I fixed a few things up, added the promise sample, and updated the PR.  Thank you :)

evmar commented on README.md:
````diff
@@ -12,35 +12,39 @@ npm install google-p12-pem
+### async/await style
+```js
+const gp12 = require('google-p12-pem');
+const pem = await gp12.getPem('/path/to/key.p12');
````
I don't think you can use 'await' at the top level, but only in an async function.

JustinBeckwith replied:
Fair point :) Updated the doc to use an async function.

evmar commented on tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+    "src/**/*.ts",
+    "test/*.ts",
+    "test/**/*.ts"
+  ],
```
I think you can leave 'include' out and it just compiles **/*.ts (obeying the exclude).

Alternatively, you can leave the exclude out.  Specifying both is redundant I think.

JustinBeckwith replied:
@ofrobots thoughts here?  This is straight from the gts auto-generated tsconfig.  It's possible there are *.ts files in the node_modules dir, so I don't think we'd want to leave that out.  Specifying the paths for *.ts files may be redundant, but it is explicit :)

ofrobots replied:
IIRC, `gts` sets things up this way because of some trouble we some trouble with `tsc` ending up looking into `node_modules` despite the `includes` section. I can't remember the details at the moment however.

I have opened https://github.com/google/ts-style/issues/70 against `gts` to figure this out.

@evmar: do you think this needs to be addressed as part of this PR?

JustinBeckwith replied:
@evmar Gentle ping :)

evmar replied:
I think you shouldn't worry about this, it's a trivial detail that is easily changed later. :)

evmar commented on src/index.ts:
```diff
@@ -0,0 +1,61 @@
+
+async function getPemAsync(filename: string) {
+  return new Promise<string>((resolve, reject) => {
+    fs.readFile(filename, {encoding: 'base64'}, (err, keyp12) => {
```
One pattern I like when using async functions is to wrap fs.readFile itself in an async function

async readFile(...): Promise<string>;

and then your code (like getPemAsync) can just

const pem = convertToPem(await readFile(filename));

This way you don't have a big Promise(....) wrapping the entirety of your logic.

(As currently written, this function is not an async function, it's just an ordinary function that returns a promise.)

JustinBeckwith replied:
I kicked around the idea of using something like pify to promisify `fs.readFile`, and then just `await`ing the promise.  Since we need to target nodejs 4/6/8, I couldn't rely on `util.promisify`.   @ofrobots thoughts?

ofrobots replied:
You could use [`pify`](npm.im/pify) or https://github.com/ljharb/util.promisify as a dependency. I do think the readability becomes a lot better if `fs.readFile` was already promisified.

JustinBeckwith replied:
Good call.  I went ahead and used pify - it is a lot cleaner 👍

JustinBeckwith commented:
@ofrobots @evmar I think I have all the feedback here covered.  If there's stuff still lingering, let me know!

ofrobots commented:
Thanks! Landed. There are very few dependent modules: https://www.npmjs.com/browse/depended/google-p12-pem with the primary one being https://www.npmjs.com/package/gtoken. It would be good to do an integration test before releasing this to npm as 1.0.0.

JustinBeckwith commented:
Rad. With the breaking change (mostly the es modules / getPem thing) I did have to make a change.  But after that, it works fine:

![image](https://user-images.githubusercontent.com/534619/32259123-91b010c6-be7b-11e7-9f50-f5d4fc84cf78.png)
